fixes #22467 - dashboard: list hosts with expired token

timogoebel opened the pull request:
![image](https://user-images.githubusercontent.com/4107560/35618207-88f815c2-067a-11e8-9d3a-7a573e5f6949.png)
![image](https://user-images.githubusercontent.com/4107560/35618235-9fb87d88-067a-11e8-9930-5b950c09da76.png)

timogoebel committed:
fixes #22467 - dashboard: list hosts in build mode

sean797 commented on app/views/dashboard/_hosts_expired_token_widget_host_list.html.erb:
```diff
@@ -0,0 +1,18 @@
+      <tr>
+        <td class="ellipsis"><%= icon_text('error-circle-o', '', kind: 'pficon') %> <%= link_to host.name, host_path(host) %></td>
+        <td class="hidden-tablet hidden-xs ellipsis"><%= host.owner %></td>
+        <td class="hidden-tablet hidden-xs ellipsis"><%= date_time_relative_value(host.created_at) %></td>
```
Would getting a value for "Built" be better then created_at? Just created_at wouldn't be relevant to hosts that have been rebuilt.

You could use Setting[:token_duration] and `host.token.expires` to work it out. Obviously that would be incorrect if someone changes Setting[:token_duration] though :(

ShimShtein commented:
Nice! Can we also add a scoped search definition for `expired?`?

dLobatog reviewed (dismissed):
Setting to WoC based on comments

ohadlevy commented on app/views/dashboard/_hosts_expired_token_widget_host_list.html.erb:
```diff
@@ -0,0 +1,18 @@
+<table class="table table-striped table-fixed">
```
should we use the table css classes helper here?

timogoebel commented:
I changed the code to use the helper and removed the created at column in the table for now. I think, we should store the time a host is set to built in the database. But I want to do that in a separate PR. I'll also create an issue to add `expired?` to scoped search. This is out of scope here (although a good idea).

Anything else to add?

This is how it looks now:
![image](https://user-images.githubusercontent.com/4107560/37201594-6f36453e-2388-11e8-8639-dd4d4665130b.png)

timogoebel commented:
@lzap actually added the required fields to show the build start time in #5292. I'll wait for the PR to get in before I continue to work on this.

lzap commented:
Nice patch, I'd also like to see list of hosts which are actually in build mode. Do you mind renaming the feature and widget to show both - there could be a filter or button to only show expired ones. I don't know how others, but my deployments have usually just few hosts in build mode.

timogoebel commented:
> I'd also like to see list of hosts which are actually in build mode.

Nice idea, I'll look into it.

timogoebel commented:
Updated. This is how it currently looks.
@lzap: Do you mind taking a look? Should be an easy review. :-)

![image](https://user-images.githubusercontent.com/4107560/41607459-5e61ad1c-73e6-11e8-8814-62485dfc953f.png)

ohadlevy commented:
@timogoebel last screenshot is very nice :) would you consider having its own page for it (vs a dashboard widget?)

timogoebel commented:
> would you consider having its own page for it (vs a dashboard widget?)

I prefer the widget for now as it's more prominent (and because I don't have time for more), but we could go crazy on a separate page (e.g. with some visual indication of what is actually happening, something like this: https://res.infoq.com/news/2016/02/netflix-spinnaker/en/resources/1spinnaker-pipelines.png). I'd really like this but I believe that's a huge amount of work to accomplish.

ekohl commented on app/views/dashboard/_hosts_in_build_mode_widget_host_list.html.erb:
```diff
@@ -0,0 +1,18 @@
+        <td class="hidden-tablet hidden-xs ellipsis"><%= build_duration(host) %></td>
+        <td class="hidden-tablet hidden-xs ellipsis"><%= date_time_relative_value(host.token.expires) %></td>
+      </tr>
+  <% end %>
```
nit: odd indenting

lzap commented on app/helpers/dashboard_helper.rb:
```diff
@@ -159,6 +159,20 @@ def out_of_sync_enabled?(origin)
+      label = _('Build in progress')
+    end
+    icon_text(icon, '', kind: 'pficon', title: label)
+  end
```
Perhaps to use different icon for expired and errored states? Just a thought.

timogoebel replied:
Patternfly doesn't have a icon that fits in my opinion.
https://www.patternfly.org/styles/icons/
Anything you like, @lzap?

ohadlevy replied:
@rohoover any suggestions?

lzap replied:
How about the bell icon? Like expired = bell = ring = end. "fa-bell" is the name.

Rohoover replied:
@lzap 

We use the bell icon for the notification drawer. I wouldn't repeat it. I'm bringing this back to the team to see if it's already been identified elsewhere. Stay tuned!

Rohoover replied:
This is what I'm proposing:

Expiring soon = fa-hourglass-end
Expired = fa-hourglass-o

timogoebel replied:
Thanks, @Rohoover good call.

lzap reviewed (dismissed):
Nice, please rebase the indenting and I am gonna test and merge thanks.

timogoebel commented:
Fixed the indent issue.

lzap commented on app/helpers/dashboard_helper.rb:
```diff
@@ -159,6 +159,20 @@ def out_of_sync_enabled?(origin)
+      label = _('Token expired')
+    elsif host.build_errors.present?
+      icon = 'error-circle-o'
+      label = _('Build errors')
```
There is one problem here - if a host has `build_errors` set, it is never in `build` mode. When I error a host by running this command:

    wget -q -O- --body-file=LICENSE --method POST --header="Content-Type: text/plain" --no-check-certificate http://localhost:3000/unattended/failed?token=689ba68a-a511-44b1-bca5-c8f3f1ff353e

then the host disappears. Therefore this branch can't be reached.

timogoebel replied:
Interestingly, I get this:

```
Encoding::UndefinedConversionError ("\xE2" from ASCII-8BIT to UTF-8):
```

Maybe we need something like this:

```ruby
.encode('utf-8', invalid: :replace, undef: :replace, replace: '_')
```

`.gitignore` works fine, our `LICENSE` does not...

lzap commented on app/views/dashboard/_hosts_in_build_mode_widget.html.erb:
```diff
@@ -0,0 +1,11 @@
+<% hosts = @data.hosts.in_build_mode.includes(:token).preload(:owner).order(created_at: :desc).limit(9) %>
```
This query must include both "in build" and also when "build errors" text is present (not empty).

lzap commented on app/views/dashboard/_hosts_in_build_mode_widget.html.erb:
```diff
@@ -0,0 +1,11 @@
+<% hosts = @data.hosts.in_build_mode.includes(:token).preload(:owner).order(created_at: :desc).limit(9) %>
+
+<h4 class="header">
+  <%= _('Hosts in Build mode') %>
```
This could be also "Hosts in Build mode or Failed hosts" or similar.

lzap reviewed (changes requested):
One remark.

timogoebel commented:
Thanks, added hosts with build errors to the widget and fixed an issue where hosts that entered build mode before this patch was applied showed an error. Also changed the icon to follox Rox' recommendation.

This is the latest screenshot:
![image](https://user-images.githubusercontent.com/4107560/41678508-8383c298-74cb-11e8-87d2-1c3c20e80cb2.png)

lzap commented:
Thank you for your contribution, merged.

alexjfisher commented on app/helpers/hosts_helper.rb:
```diff
@@ -251,7 +251,7 @@ def name_field(host)
-    return _("N/A") if host.initiated_at.nil? && host.installed_at.nil?
+    return _('N/A') if host.initiated_at.nil? || host.installed_at.nil?
     if host.installed_at.nil?
       time_ago_in_words(host.initiated_at, include_seconds: true) + " (in progress)"
```
This code is now unreachable?

timogoebel replied:
Looks unreachable, yes.

lzap replied:
Bite me, but I do see a path: when `host.installed_at.nil?` is true?!? Explain please.

alexjfisher replied:
@lzap Function will return on line 254 if host.installed_at.nil? is true.

lzap replied:
I see.

lzap replied:
I will take a look next week. My GH backlog is not getting slimmer.

lzap replied:
Solved in https://github.com/theforeman/foreman/pull/6603
